### PR TITLE
[Node Mgr] Suppot Node-MQ Topic Mapping

### PR DESCRIPTION
**Pull request by VanderChen**
Implement  #490  as Port Manager style and there still be a problem to solve.

### CODE REVIEW

File: services/data_plane_manager/src/main/java/com/futurewei/alcor/dataplane/cache/LocalCacheImpl.java [hunk 1]
```diff
@@ -135,4 +139,24 @@ public void updateLocalCache(NetworkConfiguration networkConfig) throws Exceptio
                 LOG.error("Update SubnetPorts failed: Unknown operation type");
         }
     }
+
+    @Override
+    public void addNodeInfo(NodeInfo nodeInfo) throws Exception {
+        nodeInfoCache.addNodeInfo(nodeInfo);
```
Comment (chenpiaoping) [inline services/data_plane_manager/src/main/java/com/futurewei/alcor/dataplane/cache/LocalCacheImpl.java:145]: Do we need to cache all the fields of NodeInfo here?

File: services/data_plane_manager/src/main/java/com/futurewei/alcor/dataplane/cache/NodeInfoCache.java [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+
+
+    @DurationStatistics
+    public NodeInfo getNodeInfo(String nodeIp) throws Exception {
+        NodeInfo nodeInfo;
+        try {
+            nodeInfo = nodeInfoCache.get(nodeIp);
+            assert nodeInfo != null;
+        } catch (Exception e) {
+            NodeInfo newNodeInfo = nodeManagerRestClient.getNodeInfo(nodeIp).getNodeInfo();
```
Comment (chenpiaoping) [inline services/data_plane_manager/src/main/java/com/futurewei/alcor/dataplane/cache/NodeInfoCache.java:33]: newNodeInfo may be null here?

File: services/data_plane_manager/src/main/java/com/futurewei/alcor/dataplane/cache/NodeInfoCache.java [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+            nodeInfoCache.put(newNodeInfo.getLocalIp(), newNodeInfo);
+            nodeInfo = newNodeInfo;
+        }
+        return nodeInfo;
+    }
+
+    @DurationStatistics
+    public synchronized void addNodeInfo(NodeInfo nodeInfo) throws Exception {
```
Comment (chenpiaoping) [inline services/data_plane_manager/src/main/java/com/futurewei/alcor/dataplane/cache/NodeInfoCache.java:41]: Do we need synchronized here?

File: services/data_plane_manager/src/main/java/com/futurewei/alcor/dataplane/client/pulsar/DataPlaneClientImpl.java [hunk 4]
```diff
@@ -45,35 +46,14 @@
...
-        List<String> groupTopics = this.getGroupTopics(hostIps);
-        for (String groupTopic: groupTopics) {
-            String multicastTopic = topicManager.getMulticastTopicByGroupTopic(groupTopic);
-            if (StringUtils.isEmpty(multicastTopic)) {
-                LOG.error("Can not find multicast topic by group topic:{}", groupTopic);
-                throw new MulticastTopicNotFound();
-            }
-
+        for (String hostIp : hostIps) {
+            String groupTopic = localCache.getNodeInfo(hostIp).getGroupTopic();
```
Comment (chenpiaoping) [inline services/data_plane_manager/src/main/java/com/futurewei/alcor/dataplane/client/pulsar/DataPlaneClientImpl.java:55]: localCache.getNodeInfo(hostIp) may rerturn null?

File: services/data_plane_manager/src/main/java/com/futurewei/alcor/dataplane/controller/NodeController.java [hunk 5]
```diff
@@ -0,0 +1,84 @@
...
+    }
+
+    @PutMapping({"/nodes", "v4/nodes"})
+    @DurationStatistics
+    public void updateNodeInfo(@RequestBody NodeInfoJson nodeInfoJson) throws Exception {
+        if (nodeInfoJson == null) {
+            throw new ParameterNullOrEmptyException("");
+        }
+        try {
+            nodeService.createNodeInfo(nodeInfoJson);
```
Comment (chenpiaoping) [inline services/data_plane_manager/src/main/java/com/futurewei/alcor/dataplane/controller/NodeController.java:49]: updateNodeInfo()?

File: services/data_plane_manager/src/main/java/com/futurewei/alcor/dataplane/controller/NodeController.java [hunk 6]
```diff
@@ -0,0 +1,84 @@
...
+    }
+
+    @DeleteMapping({"/nodes", "v4/nodes"})
+    @DurationStatistics
+    public void deleteNodeInfo(@RequestBody NodeInfoJson nodeInfoJson) throws Exception {
+        if (nodeInfoJson == null) {
+            throw new ParameterNullOrEmptyException("");
+        }
+        try {
+            nodeService.createNodeInfo(nodeInfoJson);
```
Comment (chenpiaoping) [inline services/data_plane_manager/src/main/java/com/futurewei/alcor/dataplane/controller/NodeController.java:63]: deleteNodeInfo()?

File: services/data_plane_manager/src/main/java/com/futurewei/alcor/dataplane/service/impl/NodeServiceImpl.java [hunk 7]
```diff
@@ -0,0 +1,40 @@
...
+    @Override
+    public void deleteNodeInfo(NodeInfoJson nodeInfoJson) throws Exception {
+        localCache.deleteNodeInfo(nodeInfoJson.getNodeInfo().getId());
+    }
+
+    @Override
+    public void createNodeInfoBulk(BulkNodeInfoJson bulkNodeInfoJson) throws Exception {
+        List<NodeInfo> nodeInfos = bulkNodeInfoJson.getNodeInfos();
+        for (NodeInfo nodeInfo : nodeInfos) {
+            localCache.addNodeInfo(nodeInfo);
```
Comment (chenpiaoping) [inline services/data_plane_manager/src/main/java/com/futurewei/alcor/dataplane/service/impl/NodeServiceImpl.java:37]: localCache should support batch operation?

File: services/data_plane_manager/src/main/java/com/futurewei/alcor/dataplane/controller/NodeController.java [hunk 8]
```diff
@@ -0,0 +1,84 @@
...
+
+    @Autowired
+    NodeService nodeService;
+
+    @PostMapping({"/nodes", "v4/nodes"})
+    @ResponseStatus(HttpStatus.CREATED)
+    @DurationStatistics
+    public void createNodeInfo(@RequestBody NodeInfoJson nodeInfoJson) throws Exception {
+        if (nodeInfoJson == null) {
+            throw new ParameterNullOrEmptyException("");
```
Comment (chenpiaoping) [inline services/data_plane_manager/src/main/java/com/futurewei/alcor/dataplane/controller/NodeController.java:32]: More checks need to be done here?

File: services/data_plane_manager/src/main/java/com/futurewei/alcor/dataplane/cache/NodeInfoCache.java (hunk 2 above)
Comment (VanderChen) [reply services/data_plane_manager/src/main/java/com/futurewei/alcor/dataplane/cache/NodeInfoCache.java]: A NodeInfoNotFound Exception has been added here.

File: services/data_plane_manager/src/main/java/com/futurewei/alcor/dataplane/cache/NodeInfoCache.java (hunk 3 above)
Comment (VanderChen) [reply services/data_plane_manager/src/main/java/com/futurewei/alcor/dataplane/cache/NodeInfoCache.java]: I have remove it.

File: services/data_plane_manager/src/main/java/com/futurewei/alcor/dataplane/client/pulsar/DataPlaneClientImpl.java (hunk 4 above)
Comment (VanderChen) [reply services/data_plane_manager/src/main/java/com/futurewei/alcor/dataplane/client/pulsar/DataPlaneClientImpl.java]: A Exception has been added here.

File: services/data_plane_manager/src/main/java/com/futurewei/alcor/dataplane/controller/NodeController.java (hunk 5 above)
Comment (VanderChen) [reply services/data_plane_manager/src/main/java/com/futurewei/alcor/dataplane/controller/NodeController.java]: I have fixed this bug.

File: services/data_plane_manager/src/main/java/com/futurewei/alcor/dataplane/controller/NodeController.java (hunk 6 above)
Comment (VanderChen) [reply services/data_plane_manager/src/main/java/com/futurewei/alcor/dataplane/controller/NodeController.java]: I have fixed this bug.

File: services/data_plane_manager/src/main/java/com/futurewei/alcor/dataplane/service/impl/NodeServiceImpl.java (hunk 7 above)
Comment (VanderChen) [reply services/data_plane_manager/src/main/java/com/futurewei/alcor/dataplane/service/impl/NodeServiceImpl.java:37]: addNodeInfoBulk method has been added to LocalCache

File: services/data_plane_manager/src/main/java/com/futurewei/alcor/dataplane/controller/NodeController.java (hunk 8 above)
Comment (VanderChen) [reply services/data_plane_manager/src/main/java/com/futurewei/alcor/dataplane/controller/NodeController.java:32]: More checks have been added here.

**Review by chenpiaoping — APPROVED** [review]

**Comment by chenpiaoping** [comment]
LGTM

**Comment by VanderChen** [comment]
@xieus Excuse me. I wonder whether this PR is ready for merge

**Review by xieus — APPROVED** [review]
LGTM.

**Pull request merged**